engineering-plans/0002-ethereum-tracing-cache.md

```diff
@@ -0,0 +1,184 @@
+# PLAN-0002: Ethereum Tracing Cache
+
+<dl>
+  <dt>Author</dt>
+  <dd>Zachary Burns</dd>
+
+  <dt>Implements</dt>
+  <dd><a href="../rfcs/0002-ethereum-tracing-cache.md">RFC-0002 Ethereum Tracing Cache</a></dd>
+
+  <dt>Engineering Plan pull request</dt>
+  <dd><a href="https://github.com/graphprotocol/rfcs/pull/9">https://github.com/graphprotocol/rfcs/pull/9</a></dd>
+
+  <dt>Obsoletes (if applicable)</dt>
+  <dd>None</dd>
+
+  <dt>Date of submission</dt>
+  <dd>2019-12-20</dd>
+
+  <dt>Date of approval</dt>
+  <dd>YYYY-MM-DD</dd>
+
+  <dt>Approved by</dt>
+  <dd>First Person, Second Person</dd>
+</dl>
+
+
+## Summary
+
+Implements RFC-0002: Ethereum Tracing Cache
+
+## Implementation
+
+These changes happen within or near `ethereum_adapter.rs`, `store.rs` and `db_schema.rs`.
+
+### Limitations
+The problem of reorg turns out to be a particularly tricky one for the cache, mostly due to ranges of blocks being requested rather than individual hashes. To sidestep this problem, only blocks that are older than the reorg threshold will be eligible for caching.
+
+This workaround has an intended side effect. The target beneficiary of the workflow involving re-sync of recently deployed subgraphs. It may be undesirable to cache subgraphs which were not recently deployed and may never be re-synced. Such a subgraph would only incur the cost of caching, but never reap the benefits of reading from the cache. By only caching blocks outside the reorg threshold, only the subgraphs which have been recently deployed and are the most likely to benefit from caching are in fact cached.
+
+### EthereumTraceCache
+
+The implementation introduces the following trait, which is implemented primarily by `Store`.
+
+```rust
+use std::ops::Range;
+struct TracesInRange {
+    range: Range<u64>,
+    traces: Vec<Trace>,
+}
+
+pub trait EthereumTraceCache: Send + Sync + 'static {
+    /// Attempts to retrieve traces from the cache. Returns ranges which were retrieved.
+    /// The results may not cover the entire range of blocks. It is up to the caller to decide
+    /// what to do with ranges of blocks that are not cached.
+    fn get_cache(contract_address: Option<H160>, blocks: Range<u64>
+        ) -> Box<dyn Future<Output=Result<Vec<TracesInRange>, Error>>>;
+    fn append_cache(contract_address: Option<H160> traces: Vec<TracesInRange>);
+}
+```
+
+#### Block schema
+
+Each cached block will exist as its own row in the database. Here is the schema for that cache:
+
+```rust
+  id -> (Nullable<Bytea>, Integer), // (contract_address, block_number)
+  trace -> Jsonb // Or Bytea
```

Comment:
In SQL, you can't have a nullable component of a composite primary key afaik. So the block number can't be part of the primary key. It would be good to have the SQL migrations in the plan.

Reply:
I'll use a multi-column index, and rename trace to traces.

Reply:
Are there any necessary migrations when just adding new tables?

Reply:
I just mean it would be nice to write out the `CREATE TABLE` commands.